storm-client/src/jvm/org/apache/storm/topology/SimpleWindowPartitionCache.java

```diff
@@ -0,0 +1,191 @@
+        boolean res = false;
+        try {
+            lock.lock();
+            if (pinned.compute(key, (k, v) -> v == null ? 0 : v - 1) <= 0) {
+                pinned.remove(key);
+                res = true;
+            }
+        } finally {
+            lock.unlock();
+        }
+        LOG.debug("pinned '{}'", pinned);
+        return res;
+    }
+
+    @Override
+    public ConcurrentMap<K, V> asMap() {
+        return map;
+    }
+
+    @Override
+    public void invalidate(K key) {
+        try {
+            lock.lock();
+            if (isPinned(key)) {
+                LOG.debug("Entry '{}' is pinned, skipping invalidation", key);
+            } else {
+                LOG.debug("Invalidating entry '{}'", key);
+                V val = map.remove(key);
+                if (val != null) {
+                    --size;
+                    pinned.remove(key);
+                    removalListener.onRemoval(key, val, RemovalCause.EXPLICIT);
+                }
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    // Get or load from the cache optionally pinning the entry
+    // so that it wont get evicted from the cache
+    private V getOrLoad(K key, boolean shouldPin) {
+        V val;
+        if (shouldPin) {
+            try {
+                lock.lock();
+                val = load(key);
+                pin(key);
+            } finally {
+                lock.unlock();
+            }
+        } else {
+            val = map.get(key);
+            if (val == null) {
+                try {
+                    lock.lock();
+                    val = load(key);
+                } finally {
+                    lock.unlock();
+                }
+            }
+        }
+
+        return val;
+    }
+
+    private V load(K key) {
+        V val = map.get(key);
+        if (val == null) {
+            val = cacheLoader.load(key);
+            if (val == null) {
+                throw new NullPointerException("Null value");
+            }
+            ensureCapacity();
+            map.put(key, val);
+            ++size;
+        }
+        return val;
+    }
+
+    private void ensureCapacity() {
+        if (size >= maximumSize) {
+            Iterator<Map.Entry<K, V>> it = map.descendingMap().entrySet().iterator();
+            while (it.hasNext()) {
+                Map.Entry<K, V> next = it.next();
+                if (!isPinned(next.getKey())) {
+                    it.remove();
+                    removalListener.onRemoval(next.getKey(), next.getValue(), RemovalCause.REPLACED);
+                    --size;
+                    break;
+                }
+            }
+        }
+    }
+
+    private void pin(K key) {
+        LOG.debug("pin '{}'", key);
+        pinned.compute(key, (k, v) -> v == null ? 1L : v + 1);
+        LOG.debug("pinned '{}'", pinned);
+    }
+
+    private boolean isPinned(K key) {
+        return pinned.getOrDefault(key, 0L) > 0;
```

Comment:
containsKey should work here since we are removing the entry in `unpin` if the value <= 0, but prefer to leave it as is to be defensive against future code refactoring.